File: exodia/bridge.lua
-- EXODIA-4B :: Left Arm (peca 1/5)
-- Ponte BizHawk -> Python. Roda DENTRO do EmuHawk.
--
-- Como usar:
--   1. Suba o servidor Python (exodia/bridge.py) PRIMEIRO.
--   2. EmuHawk.exe --socket_ip=127.0.0.1 --socket_port=55355 --lua=<este arquivo> <rom>
--
-- Protocolo: o Python e o SERVIDOR, o BizHawk e o CLIENTE.
-- A cada iteracao o Lua manda o resultado do comando anterior e recebe o proximo.
-- O Python DEVE prefixar suas respostas com "<tamanho> " (exigencia do BizHawk >= 2.6.2).

-- Quanto o Lua espera por um comando antes de desistir e adiantar um frame.
--
-- Era 50 ms, e isso fazia o jogo ANDAR SOZINHO enquanto o modelo pensava: uma
-- inferencia de 30 s virava ~600 frames de jogo rodando sem ninguem no
-- controle. O agente decidia sobre um estado e agia sobre outro.
--
-- Com 2 s o emulador fica praticamente congelado durante a inferencia - que e
-- justamente a premissa do projeto: o jogo e por turnos e espera por nos, entao
-- latencia custa tempo de relogio, nunca qualidade de jogo. E de quebra o
-- BizHawk para de disputar CPU com o Ollama.
--
-- Nao e infinito de proposito: se o processo Python morrer, o EmuHawk volta a
-- responder em 2 s e da para fechar a janela na mao.
local POLL_TIMEOUT_MS = 2000

comm.socketServerSetTimeout(POLL_TIMEOUT_MS)

-- ---------------------------------------------------------------- utilidades

local function hexdump(bytes)
  local t = {}
  for i = 1, #bytes do t[i] = string.format("%02X", bytes[i]) end
  return table.concat(t)
end

local function split(s)
  local t = {}
  for w in string.gmatch(s, "%S+") do t[#t + 1] = w end
  return t
end

-- Le n bytes crus a partir de addr no dominio informado (default MainRAM).
-- IMPORTANTE: mascara 0x1FFFFF converte endereco PS1 (0x800xxxxx) em offset MainRAM.
local function read_range(addr, len, domain)
  domain = domain or "MainRAM"
  local out = {}
  for i = 0, len - 1 do
    out[i + 1] = memory.read_u8(addr + i, domain)
  end
  return out
end

-- ---------------------------------------------------------------- comandos

local BUTTONS = {}  -- cache do nome do controle

local function press(button, frames)
  frames = frames or 2
  for _ = 1, frames do
    local t = {}
    t[button] = true
    joypad.set(t, 1)
    emu.frameadvance()
  end
  -- solta e deixa 1 frame limpo
  emu.frameadvance()
  return "OK"
end

local handlers = {}

handlers.PING = function() return "PONG" end

-- BUTTONS -> nomes EXATOS dos botoes aceitos por joypad.set neste core,
-- cada um acompanhado dos seus bytes em hex.
-- Nunca adivinhe: no core PSX o D-pad e "D-Pad Right" (nao "Right") e o
-- botao de confirmar e "X" (nao "Cross"). Circulo/quadrado/triangulo sao
-- simbolos Unicode, por isso o hex - evita qualquer duvida de encoding.
local function tohex(s)
  return (s:gsub(".", function(c) return string.format("%02X", c:byte()) end))
end

handlers.BUTTONS = function()
  local t = {}
  for name, _ in pairs(joypad.get(1)) do
    t[#t + 1] = tohex(name) .. ":" .. name
  end
  table.sort(t)
  return table.concat(t, ",")
end

local function fromhex(h)
  return (h:gsub("%x%x", function(cc)
    return string.char(tonumber(cc, 16))
  end))
end

-- PRESSHEX <hex_do_nome> [frames] -> aperta o botao endereçado por bytes
handlers.PRESSHEX = function(a)
  local name = fromhex(a[2] or "")
  if name == "" then return "ERR hex vazio" end
  return press(name, tonumber(a[3]))
end

handlers.INFO = function()
  local parts = {
    "system=" .. tostring(emu.getsystemid()),
    "core=" .. tostring(emu.getdisplaytype and emu.getdisplaytype() or "?"),
    "frame=" .. tostring(emu.framecount()),
    "rom=" .. tostring(gameinfo.getromname()),
    "hash=" .. tostring(gameinfo.getromhash()),
  }
  return table.concat(parts, "|")
end

-- READ <hex_addr> <nbytes> [domain]  -> valor sem sinal, little-endian
handlers.READ = function(a)
  local addr = tonumber(a[2], 16) & 0x1FFFFF
  local n = tonumber(a[3]) or 1
  local domain = a[4] or "MainRAM"
  local v
  if n == 1 then v = memory.read_u8(addr, domain)
  elseif n == 2 then v = memory.read_u16_le(addr, domain)
  elseif n == 4 then v = memory.read_u32_le(addr, domain)
  else return "ERR tamanho deve ser 1, 2 ou 4" end
  return tostring(v)
end

-- READRANGE <hex_addr> <len> [domain] -> hex string
handlers.READRANGE = function(a)
  local addr = tonumber(a[2], 16) & 0x1FFFFF
  local len = tonumber(a[3]) or 1
  local domain = a[4] or "MainRAM"
  return hexdump(read_range(addr, len, domain))
end

-- DOMAINS -> lista os dominios de memoria disponiveis no core atual
-- ATENCAO: o BizHawk devolve esse array INDEXADO EM 0. Usar ipairs() aqui
-- silenciosamente pula a primeira entrada - que e justamente a MainRAM.
handlers.DOMAINS = function()
  local list = memory.getmemorydomainlist()
  local t = {}
  local i = 0
  while list[i] ~= nil do
    local d = list[i]
    t[#t + 1] = d .. ":" .. tostring(memory.getmemorydomainsize(d))
    i = i + 1
  end
  -- fallback caso alguma versao devolva indexado em 1
  if #t == 0 then
    for _, d in ipairs(list) do
      t[#t + 1] = d .. ":" .. tostring(memory.getmemorydomainsize(d))
    end
  end
  return table.concat(t, ",")
end

-- PRESS <button> [frames]
handlers.PRESS = function(a)
  return press(a[2], tonumber(a[3]))
end

-- SEQ <btn:frames,btn:frames,...>
handlers.SEQ = function(a)
  for item in string.gmatch(a[2] or "", "[^,]+") do
    local btn, fr = string.match(item, "([^:]+):?(%d*)")
    press(btn, tonumber(fr) or 2)
  end
  return "OK"
end

-- FRAME [n] -> avanca n frames
handlers.FRAME = function(a)
  local n = tonumber(a[2]) or 1
  for _ = 1, n do emu.frameadvance() end
  return tostring(emu.framecount())
end

-- FREERUN <frames> -> roda o jogo em velocidade normal por N frames e devolve
-- os botoes que o HUMANO apertou, com o frame de cada aperto.
--
-- Existe para observar uma pessoa jogando: o laco normal da ponte so avanca um
-- frame por poll do socket, o que deixa o jogo lento demais para alguem jogar.
-- Aqui o Lua fica no controle do avanco e so devolve o log no fim.
handlers.FREERUN = function(a)
  local n = tonumber(a[2]) or 600
  local log = {}
  local anterior = ""
  for _ = 1, n do
    local apertados = {}
    for nome, ligado in pairs(joypad.get(1)) do
      if ligado == true then apertados[#apertados + 1] = tohex(nome) end
    end
    table.sort(apertados)
    local atual = table.concat(apertados, "+")
    if atual ~= anterior and atual ~= "" then
      log[#log + 1] = emu.framecount() .. "=" .. atual
    end
    anterior = atual
    emu.frameadvance()
  end
  return table.concat(log, ",")
end

-- SCREENSHOT <path>
handlers.SCREENSHOT = function(a)
  local path = a[2]
  if not path then return "ERR falta o caminho" end
  client.screenshot(path)
  return "OK " .. path
end

-- SPEED <percent>  (100 = normal; valores altos = fast-forward)
handlers.SPEED = function(a)
  local pct = tonumber(a[2]) or 100
  client.speedmode(pct)
  return "OK " .. tostring(pct)
end

-- SAVESTATE <path> / LOADSTATE <path>
handlers.SAVESTATE = function(a)
  savestate.save(a[2]); return "OK"
end
handlers.LOADSTATE = function(a)
  savestate.load(a[2]); return "OK"
end

-- MMFDUMP <hex_addr> <len> <mmf_name> -> despeja bloco grande via memory-mapped file
handlers.MMFDUMP = function(a)
  local addr = tonumber(a[2], 16) & 0x1FFFFF
  local len = tonumber(a[3])
  local name = a[4] or "exodia_ram"
  local n = comm.mmfCopyFromMemory(name, addr, len, "MainRAM")
  return "OK " .. tostring(n)
end

-- ---------------------------------------------------------------- loop

local function handle(line)
  if not line or line == "" then return nil end
  local a = split(line)
  local h = handlers[string.upper(a[1] or "")]
  if not h then return "ERR comando desconhecido: " .. tostring(a[1]) end
  local ok, res = pcall(h, a)
  if not ok then return "ERR " .. tostring(res) end
  return res
end

console.log("EXODIA bridge.lua carregado. Aguardando o servidor Python...")

local pending = "HELLO"

while true do
  comm.socketServerSend(pending)
  local cmd = comm.socketServerResponse()
  if cmd and cmd ~= "" then
    local res = handle(cmd)
    pending = res or "OK"
  else
    -- nada chegou: adianta um frame para a UI do EmuHawk nao congelar
    pending = "IDLE"
    emu.frameadvance()
  end
end

File: exodia/bridge.py
"""EXODIA-4B :: Left Arm (peca 1/5)

Servidor TCP que conversa com o bridge.lua rodando dentro do BizHawk.

Direcao do protocolo (detalhe que todo mundo erra):
    o Lua do BizHawk so tem CLIENTE de socket, entao o PYTHON e o SERVIDOR.

Formato do enquadramento:
    BizHawk >= 2.6.2 exige que TUDO que o servidor manda venha prefixado com
    "<tamanho_em_decimal> ". O que o BizHawk manda de volta tambem vem
    prefixado, entao o parser aceita as duas formas por seguranca.

Uso tipico:
    with Bridge() as b:
        b.start()                 # espera o EmuHawk conectar
        lp = b.read_u16(0x800EA004)
"""

from __future__ import annotations

import socket
import threading
from contextlib import suppress

DEFAULT_HOST = "127.0.0.1"
DEFAULT_PORT = 55355

# Mascara que converte endereco do espaco PS1 (0x800xxxxx) em offset da MainRAM.
PSX_RAM_MASK = 0x1FFFFF


class BridgeError(RuntimeError):
    pass


class Bridge:
    """Servidor TCP com semantica de request/response contra o bridge.lua.

    O Lua roda um loop: manda o resultado do comando anterior, entao espera o
    proximo comando. Portanto cada `command()` faz: recebe o status pendente,
    manda o comando, recebe o resultado.
    """

    def __init__(self, host: str = DEFAULT_HOST, port: int = DEFAULT_PORT,
                 timeout: float = 30.0) -> None:
        self.host = host
        self.port = port
        self.timeout = timeout
        self._srv: socket.socket | None = None
        self._conn: socket.socket | None = None
        self._buf = b""
        self._lock = threading.Lock()

    # ------------------------------------------------------------ ciclo de vida

    def listen(self) -> None:
        self._srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._srv.bind((self.host, self.port))
        self._srv.listen(1)

    def accept(self, timeout: float = 120.0) -> None:
        if self._srv is None:
            self.listen()
        assert self._srv is not None
        self._srv.settimeout(timeout)
        try:
            self._conn, _ = self._srv.accept()
        except socket.timeout as exc:
            raise BridgeError(
                f"o EmuHawk nao conectou em {timeout}s. Ele foi aberto com "
                f"--socket_ip={self.host} --socket_port={self.port} e com o "
                f"--lua=bridge.lua?"
            ) from exc
        self._conn.settimeout(self.timeout)

    def start(self, timeout: float = 120.0) -> str:
        """Sobe o servidor, espera a conexao e consome o HELLO inicial."""
        self.listen()
        return self.start_after_listen(timeout)

    def start_after_listen(self, timeout: float = 120.0) -> str:
        """Como start(), mas para quem ja chamou listen() antes.

        Util quando o processo precisa estar ouvindo ANTES de lancar o EmuHawk,
        senao o cliente Lua nao acha o servidor.
        """
        self.accept(timeout)
        return self._recv_message()

    def close(self) -> None:
        for s in (self._conn, self._srv):
            if s is not None:
                with suppress(OSError):
                    s.close()
        self._conn = self._srv = None

    def __enter__(self) -> "Bridge":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # ------------------------------------------------------- enquadramento

    def _recv_message(self) -> str:
        """Le uma mensagem. Aceita com ou sem prefixo de tamanho."""
        assert self._conn is not None
        # tenta interpretar "<n> <payload>"
        while True:
            sp = self._buf.find(b" ")
            if sp != -1:
                head = self._buf[:sp]
                if head.isdigit():
                    n = int(head)
                    if len(self._buf) >= sp + 1 + n:
                        msg = self._buf[sp + 1: sp + 1 + n]
                        self._buf = self._buf[sp + 1 + n:]
                        return msg.decode("utf-8", "replace")
                else:
                    # sem prefixo: devolve o que houver ate agora
                    data, self._buf = self._buf, b""
                    return data.decode("utf-8", "replace").strip()
            chunk = self._conn.recv(65536)
            if not chunk:
                raise BridgeError("o EmuHawk fechou a conexao")
            self._buf += chunk

    def _send_message(self, msg: str) -> None:
        assert self._conn is not None
        payload = msg.encode("utf-8")
        self._conn.sendall(f"{len(payload)} ".encode("ascii") + payload)

    # --------------------------------------------------------------- comandos

    def command(self, cmd: str) -> str:
        """Manda um comando e devolve o resultado."""
        if self._conn is None:
            raise BridgeError("nao conectado; chame start() antes")
        with self._lock:
            self._send_message(cmd)
            while True:
                res = self._recv_message()
                # o Lua manda IDLE quando um poll expira sem comando: ignora
                if res != "IDLE":
                    return res

    # --------------------------------------------------- helpers de alto nivel

    def ping(self) -> bool:
        return self.command("PING") == "PONG"

    def info(self) -> dict[str, str]:
        raw = self.command("INFO")
        out: dict[str, str] = {}
        for part in raw.split("|"):
            if "=" in part:
                k, v = part.split("=", 1)
                out[k] = v
        return out

    def domains(self) -> dict[str, int]:
        raw = self.command("DOMAINS")
        out: dict[str, int] = {}
        for part in raw.split(","):
            if ":" in part:
                k, v = part.rsplit(":", 1)
                with suppress(ValueError):
                    out[k] = int(v)
        return out

    def main_ram(self) -> str:
        """Descobre o nome do dominio de RAM principal (2 MB) neste core.

        O nome varia entre os cores de PSX do BizHawk (Nymashock x Octoshock),
        entao nunca assuma "MainRAM" - detecte.
        """
        doms = self.domains()
        for name, size in doms.items():
            if size == 2 * 1024 * 1024:
                return name
        # fallback: o maior dominio gravavel que nao seja video/audio/bios
        ignore = ("GPURAM", "SPURAM", "BiosROM", "DCache", "Memcard", "Waterbox")
        cand = {k: v for k, v in doms.items()
                if v > 0 and not any(k.startswith(i) for i in ignore)}
        if not cand:
            raise BridgeError(f"nenhum dominio de RAM plausivel em: {doms}")
        return max(cand, key=cand.get)

    def _read(self, addr: int, size: int, domain: str = "MainRAM") -> int:
        res = self.command(f"READ {addr & PSX_RAM_MASK:06X} {size} {domain}")
        if res.startswith("ERR"):
            raise BridgeError(res)
        return int(res)

    def read_u8(self, addr: int, domain: str = "MainRAM") -> int:
        return self._read(addr, 1, domain)

    def read_u16(self, addr: int, domain: str = "MainRAM") -> int:
        return self._read(addr, 2, domain)

    def read_u32(self, addr: int, domain: str = "MainRAM") -> int:
        return self._read(addr, 4, domain)

    def read_bytes(self, addr: int, length: int, domain: str = "MainRAM") -> bytes:
        res = self.command(f"READRANGE {addr & PSX_RAM_MASK:06X} {length} {domain}")
        if res.startswith("ERR"):
            raise BridgeError(res)
        return bytes.fromhex(res)

    # ------------------------------------------------------------- botoes
    #
    # ARMADILHA: no core PSX do BizHawk o D-pad se chama "D-Pad Right" (nao
    # "Right") e o confirmar se chama "X" (nao "Cross"). Circulo, quadrado e
    # triangulo sao simbolos Unicode. Por isso resolvemos tudo pelos BYTES
    # reais que o core reporta, e nunca por nome adivinhado.

    _ALIASES = {
        "up": ("D-Pad Up",),
        "down": ("D-Pad Down",),
        "left": ("D-Pad Left",),
        "right": ("D-Pad Right",),
        "cross": ("X",),
        "start": ("Start",),
        "select": ("Select",),
        "l1": ("L1",), "l2": ("L2",), "r1": ("R1",), "r2": ("R2",),
        # simbolos: casados por codepoint, nao por texto digitado
        "circle": ("○", "◯", "⭕", "O"),
        "square": ("□", "⬜"),
        "triangle": ("△", "▵"),
    }

    def _button_map(self) -> dict[str, str]:
        """alias amigavel -> hex dos bytes exatos do nome do botao."""
        if getattr(self, "_btnmap", None):
            return self._btnmap  # type: ignore[return-value]
        raw = self.command("BUTTONS")
        by_name: dict[str, str] = {}
        for item in raw.split(","):
            if ":" in item:
                hx, name = item.split(":", 1)
                by_name[name] = hx
        out: dict[str, str] = {}
        for alias, cands in self._ALIASES.items():
            for c in cands:
                if c in by_name:
                    out[alias] = by_name[c]
                    break
        self._btnmap = out  # type: ignore[attr-defined]
        return out

    def buttons(self) -> dict[str, str]:
        """Aliases disponiveis neste core (para diagnostico)."""
        return dict(self._button_map())

    def press(self, button: str, frames: int = 2) -> None:
        """Aperta um botao. Aceita alias amigavel ('right', 'cross') ou o
        nome exato do core."""
        hx = self._button_map().get(button.strip().lower())
        if hx:
            self.command(f"PRESSHEX {hx} {frames}")
        else:
            self.command(f"PRESS {button} {frames}")

    def sequence(self, seq: str) -> None:
        """seq no formato 'down:2,down:2,cross:2'."""
        for item in seq.split(","):
            item = item.strip()
            if not item:
                continue
            btn, _, fr = item.partition(":")
            self.press(btn, int(fr) if fr.isdigit() else 2)

    def frame_advance(self, n: int = 1) -> int:
        return int(self.command(f"FRAME {n}"))

    def screenshot(self, path: str) -> str:
        return self.command(f"SCREENSHOT {path}")

    def speed(self, percent: int) -> None:
        self.command(f"SPEED {percent}")

    def savestate(self, path: str) -> None:
        self.command(f"SAVESTATE {path}")

    def loadstate(self, path: str) -> None:
        self.command(f"LOADSTATE {path}")
